### PR TITLE
Bld reason count all phases

### PR DESCRIPTION
**Pull request by gabemontero**
@openshift/sig-developer-experience ptal

this PR is in reference to our work for candidate metrics / alerts for builds this sprint (https://trello.com/c/kl4dzl8F/1483-2-define-candidate-metricsalerts-for-builds)

Aside from functionality changes, there are some questions to generate discussion.

As you review, please do so with this context:
- this stemmed from recent starter us east debug attempts based on current metrics
- this is very much an initial iteration ... while I think the core of these changes is valid and needed, there is certainly potential, additional changes down the line wrt metric types, label/dimensions as well, that could be warranted; I foresee multiple PRs perhaps to progress vs. a prolonged iterative cycle with this change, but let's see if folks get on board with that or not

### CODE REVIEW

**Comment by gabemontero** [comment]
[(2) Define candidate metrics+alerts for builds](https://trello.com/c/kl4dzl8F/1483-2-define-candidate-metricsalerts-for-builds)

File: pkg/build/apis/build/types.go [hunk 1]
```diff
@@ -463,7 +463,7 @@ const (
 
 	// StatusReasonExceededRetryTimeout is an error condition when the build has
 	// not completed and retrying the build times out.
-	StatusReasonExceededRetryTimeout StatusReason = "ExceededRetryTimeout"
+	StatusReasonExceededRetryTimeout StatusReason = "ExceededRetryTimeout" //TODO not used currently, api or v1
```
Comment (bparees) [inline pkg/build/apis/build/types.go]: not sure what the todo(action) would be?  probably worthy of note but not a todo (we can't remove them since they are apis and there might still be builds out there somewhere that have these values)

File: pkg/build/controller/build/build_controller.go [hunk 2]
```diff
@@ -414,6 +414,7 @@ func (bc *BuildController) cancelBuild(build *buildapi.Build) (*buildUpdate, err
 	podName := buildapi.GetBuildPodName(build)
 	err := bc.podClient.Pods(build.Namespace).Delete(podName, &metav1.DeleteOptions{})
 	if err != nil && !errors.IsNotFound(err) {
+		//TODO - buildapi.StatusReasonCancelBuildFailed is not currently used ... should it be used here?
```
Comment (bparees) [inline pkg/build/controller/build/build_controller.go]: yeah probably.

File: pkg/build/metrics/prometheus/metrics.go [hunk 3]
```diff
@@ -77,72 +77,79 @@ func (bc *buildCollector) Collect(ch chan<- prometheus.Metric) {
...
 	for _, b := range result {
-		cc, cp, ec, nc, pc, rc, r := bc.collectBuild(ch, b)
-		for key, value := range r {
-			reasons[key] = reasons[key] + value
+		r := bc.collectBuild(ch, b)
+		for phase, reasonMap := range r {
+			currentPhaseMap := counts[phase]
+			for reason, count := range reasonMap {
+				if _, ok := currentPhaseMap[reason]; !ok {
+					currentPhaseMap[reason] = 0
```
Comment (jim-minter) [inline pkg/build/metrics/prometheus/metrics.go]: shouldn't be necessary - reading the map when the key doesn't exist will return the zero value.

File: pkg/build/metrics/prometheus/metrics.go [hunk 4]
```diff
@@ -77,72 +77,79 @@ func (bc *buildCollector) Collect(ch chan<- prometheus.Metric) {
 		return
 	}
 
-	// since we do not collect terminal build metrics on a per build basis, collectBuild will return counts
-	// to be added to the total amount posted to prometheus
-	var cancelledCount, completeCount, errorCount, newCount, pendingCount, runningCount int
-	reasons := map[string]int{}
+	// collectBuild will return counts for the build's phase/reason tuple,
+	// and counts for these tuples be added to the total amount posted to prometheus
+	counts := map[string]map[string]int{
```
Comment (jim-minter) [inline pkg/build/metrics/prometheus/metrics.go]: I think defining a `type foo struct { phase, reason string }` and a single-level `map[foo]int` would be clearer

File: pkg/build/metrics/prometheus/metrics.go (hunk 3 above)
Comment (gabemontero) [reply pkg/build/metrics/prometheus/metrics.go]: I had wondered about that given how it was but couldn't remember and then forgot to look and potentially change 

Thanks for the heads up - will change in next push

File: pkg/build/metrics/prometheus/metrics.go (hunk 4 above)
Comment (gabemontero) [reply pkg/build/metrics/prometheus/metrics.go]: agreed / like that ... part of next push

File: pkg/build/metrics/prometheus/metrics.go [hunk 5]
```diff
@@ -77,72 +77,79 @@ func (bc *buildCollector) Collect(ch chan<- prometheus.Metric) {
...
-	if len(reasons) == 0 {
-		addCountGauge(ch, buildCountDesc, failedPhase, "", float64(0))
 	}
-	for reason, count := range reasons {
-		addCountGauge(ch, buildCountDesc, failedPhase, reason, float64(count))
+
+	for phase, reasonMap := range counts {
+		// explicitly note there are no items of this phase
+		// we do not worry about posting 0 for a particular reason of a phase
+		if len(reasonMap) == 0 {
```
Comment (jim-minter) [inline pkg/build/metrics/prometheus/metrics.go]: why is this necessary, out of interest?

**Comment by jim-minter** [comment]
@gabemontero I think it'd be good if we could compare notes on the changes I think we should be making to the template prometheus metrics before this merges.

File: pkg/build/metrics/prometheus/metrics.go (hunk 5 above)
Comment (gabemontero) [reply pkg/build/metrics/prometheus/metrics.go]: necessary, no

but in experimenting with the console, I found it more intuitive when I could explicitly see that there were say 0 failed builds (vs. there simply being an absence of metrics with the label failed)

with the build polling, if there are no failed builds, etc, no metrics get registered

**Comment by gabemontero** [comment]
absolutely @jim-minter - my disclaimer around "additional changes down the line wrt metric types .." stemmed from your "Candidate metrics and alerts for Templates / TemplateInstance controller / Template process endpoint" document and accompanying email.

and as I hinted at as well, whether those discussion mutate this PR further, result in subsequent PRs this sprint .... we'll see after those discussions.

we'll sort out logistics, at least start to talk to it, during Thursday scrum

File: pkg/build/metrics/prometheus/metrics.go (hunk 5 above)
Comment (gabemontero) [reply pkg/build/metrics/prometheus/metrics.go]: that said ... i'm fine with removing, and having consumers use `count(...) == 0` to get the equivalent

File: pkg/build/apis/build/types.go (hunk 1 above)
Comment (gabemontero) [reply pkg/build/apis/build/types.go]: comment will be switched to a simple note/fyi

File: pkg/build/controller/build/build_controller.go (hunk 2 above)
Comment (gabemontero) [reply pkg/build/controller/build/build_controller.go]: actually in revisiting usage of `transitionToPhase` I'm not so sure ... it would change the behavior as to when the error was returned

i'm going to remove the comment, not rock the boat here

**Comment by gabemontero** [comment]
updates pushed

**Comment by gabemontero** [comment]
Interesting ... there is a new integration test for metrics (at least new in that it wasn't around when I last played with this) that leverages `openshift_build_total`, and it seems to be failing.

Don't know why it would be, but I'll try running locally, see what is up.

**Comment by bparees** [comment]
one nit and lgtm (outside the integration test issue you are looking into)

**Comment by gabemontero** [comment]
Figured out the integration test failure - it is a result of the change of not adding metrics even if there are no builds of a certain type, or in the integration test in particular, no builds at all.

I see the test fail locally, and then if I temporarily add the "explicitly say there are 0 completed, 0 failed, etc. that the test passes.

I'm on the fence at the moment as to whether to simple remove the openshift_build_total from the integration test (we do have the extended test which validates something very similar), or return to explicitly citing there are 0 builds for the phases.

**Comment by bparees** [comment]
> return to explicitly citing there are 0 builds for the phases.

personally i don't have a problem w/ explicitly reporting 0 for phases that have no builds in that phase.  it does provide a hint to the metric consumer that the metric exists and might be worth querying on.

**Comment by gabemontero** [comment]
Thinking about it some more ... it actually might be confusing to have phase foo, reason "" at 0, and phase foo reason bar at say 7.

Going to stand pat with this new approach, update the integration test

**Comment by gabemontero** [comment]
update pushed / commits squashed / do no merge wip label removed

itgm @bparees if still OK with things

File: pkg/build/apis/build/types.go [hunk 6]
```diff
@@ -435,7 +435,7 @@ type StatusReason string
 // These are the valid reasons of build statuses.
 const (
 	// StatusReasonError is a generic reason for a build error condition.
-	StatusReasonError StatusReason = "Error"
+	StatusReasonError StatusReason = "Error" //TODO not used currently, api or v1
```
Comment (bparees) [inline pkg/build/apis/build/types.go]: not a todo

**Comment by gabemontero** [comment]
On Thu, Feb 15, 2018 at 4:39 PM, Ben Parees <notifications@github.com>
wrote:

> *@bparees* commented on this pull request.
> ------------------------------
>
> In pkg/build/apis/build/types.go
> <https://github.com/openshift/origin/pull/18619#discussion_r168616298>:
>
> > @@ -435,7 +435,7 @@ type StatusReason string
>  // These are the valid reasons of build statuses.
>  const (
>  	// StatusReasonError is a generic reason for a build error condition.
> -	StatusReasonError StatusReason = "Error"
> +	StatusReasonError StatusReason = "Error" //TODO not used currently, api or v1
>
> not a todo
>
ah thought I fixed that

**Comment by bparees** [comment]
fix that last comment and lgtm

**Comment by gabemontero** [comment]
TODO removed - thanks
